### DISCUSSION TITLE
PrimaryKeyConstraint: Argument 'postgresql_include' is not accepted by dialect 'postgresql'PrimaryKeyConstraint

### DISCUSSION THREAD

**Discussion by LeResKP** [sqlalchemy/sqlalchemy#10664] [Usage Questions]
When trying to create a PrimaryKeyConstraint postgresql_include is not supported. I checked a bit the code and it seems there is no postgres specific kw allowed for PrimaryKeyConstraint.

This example use alambic, but it's the same when creating a Table().

```
    op.create_table(
        "user",
        sa.Column("iduser", sa.Integer(), autoincrement=False, nullable=False),
        sa.Column("value", sa.Integer(), nullable=False),
        sa.PrimaryKeyConstraint(
            "iduser",
            name="brackethistory_iduser_idproduct_idbracket_date_incl_value_pkey",
            postgresql_include=["value"],
        ),
    )

```

Here is the error traceback:
```
  File "/home/aurelien/a9engine/alembic/versions/f0eefd074216_auto.py", line 392, in upgrade
    sa.PrimaryKeyConstraint(
  File "/home/aurelien/venvs/a9engine/lib64/python3.9/site-packages/sqlalchemy/sql/schema.py", line 3839, in __init__
    super(PrimaryKeyConstraint, self).__init__(*columns, **kw)
  File "/home/aurelien/venvs/a9engine/lib64/python3.9/site-packages/sqlalchemy/sql/schema.py", line 3285, in __init__
    Constraint.__init__(self, **kw)
  File "/home/aurelien/venvs/a9engine/lib64/python3.9/site-packages/sqlalchemy/sql/schema.py", line 3129, in __init__
    self._validate_dialect_kwargs(dialect_kw)
  File "/home/aurelien/venvs/a9engine/lib64/python3.9/site-packages/sqlalchemy/sql/base.py", line 428, in _validate_dialect_kwargs
    raise exc.ArgumentError(
sqlalchemy.exc.ArgumentError: Argument 'postgresql_include' is not accepted by dialect 'postgresql' on behalf of <class 'sqlalchemy.sql.schema.PrimaryKeyConstraint'>
```

When creating an index, it works.
```
    op.create_index(
        "idx_bracketoftest_idbracket", "bracketoftest", ["idbracket"], unique=False, postgresql_include=["idtest"]
   ``` 

Is there a reason this is not working? Or it's a missing feature ?

Thanks in advance for your help

**Comment by zzzeek** [comment]
PG adds new features constantly, so apparently per https://www.postgresql.org/docs/current/sql-createtable.html INCLUDE is accepted in PRIMARY KEY, UNIQUE and EXCLUDE now , not clear when this was added.    SQLAlchemy's constructs would require modifications to support these directly, for all three.

**Reply by zzzeek** [reply]
cc @CaselIT as this would impact alembic, again (for UNIQUE only right now).

**Reply by CaselIT** [reply]
> not clear when this was added

doc seem to indicate 11:
- present here https://www.postgresql.org/docs/11/sql-createtable.html
- missing here https://www.postgresql.org/docs/10/sql-createtable.html

**Comment by CaselIT** [comment]
> Is there a reason this is not working? Or it's a missing feature ?

I also noticed it earlier this year, but worked around it by using an index. This is just a missing feature

Gonna open an issue for it. This is a bit more involved since also reflection would need to be supported

**Reply by LeResKP** [reply]
Thanks for your quick answers guys.

@CaselIT Do you remember how you did the work around using the index ?

**Reply by CaselIT** [reply]
I used an index instead of a primary key

**Comment by decaz** [comment]
Hope you can backport this feature in 1.4 branch 🙏🏻

**Reply by zzzeek** [reply]
we're into 2.1 dev already and this is a big feature add so it's unlikely to be in 2.0 even
